packages/react/src/I18nProvider.tsx

```diff
@@ -25,6 +25,34 @@ export function useLingui(): I18nContext {
   return context
 }
 
+export function withI18n(o: Object = {}): <P extends Object>(WrappedComponent: React.FunctionComponent<P>) => React.ComponentClass<P> {
+  return <P extends Object>(WrappedComponent: React.FunctionComponent<P>) => {
+    if (process.env.NODE_ENV !== "production") {
+      if (typeof o === "function" || React.isValidElement(o)) {
+        throw new Error(
+          "withI18n([options]) takes options as a first argument, " +
+            "but received React component itself. Without options, the Component " +
+            "should be wrapped as withI18n()(Component), not withI18n(Component)."
+        )
+      }
+    }
+
+    return class extends React.Component<P> {
```

Comment:
Why use a class here? You can just go with `useLingui()` hook and functional component. That also means you can drop that check for Provider.

Reply:
You're absolutly right Fredy, i'll change it asap, my mind was thinking hoc===classes, and that's why i've used a class

Reply:
Originally I though about:

```js
function withI18n(Component) {
  return props => {
    const { i18n } = useLingui()
    return <Component {...props} i18n={i18n} />
  }
}
```

but then I saw your comment about refs.. Afterall, function components can't have refs (unless you wrap then in `React.forwardRef`).

But we can keep it simple and add support for refs as soon as someone raises the issue.

Anyway, thanks both for the input!

Reply:
Done guys, refactored to functional component! Keeping the error advise if people use it like for me it's a best practice and will avoid a lot of issues of people asking why the withI18n doesnt work:

```
withI18n(Component)
```

instead of:
```
withI18n()(Component)
```

Reply:
I'm not sure why tests are failing, on my machine
Test Suites: 1 skipped, 24 passed, 24 of 25 total
Tests:       15 skipped, 212 passed, 227 total

Reply:
Tests fixed! :shipit: